runtime/near-vm-logic/src/logic.rs

```diff
@@ -0,0 +1,1010 @@
+        offset: u64,
+        num_elements: u64,
+    ) -> Result<Vec<u64>> {
+        let memory_len = num_elements
+            .checked_mul(size_of::<u64>() as u64)
+            .ok_or(HostError::MemoryAccessViolation)?;
+        let data = Self::memory_get(memory, offset, memory_len)?;
+        Ok(data
+            .chunks(size_of::<u64>())
+            .map(|buf| {
+                assert_eq!(buf.len(), size_of::<u64>());
+                let mut array = [0u8; size_of::<u64>()];
+                array.copy_from_slice(buf);
+                u64::from_le_bytes(array)
+            })
+            .collect())
+    }
+
+    // #################
+    // # Registers API #
+    // #################
+
+    /// Writes the entire content from the register `register_id` into the memory of the guest starting with `ptr`.
+    ///
+    /// # Arguments
+    ///
+    /// * `register_id` -- a register id from where to read the data;
+    /// * `ptr` -- location on guest memory where to copy the data.
+    ///
+    /// # Errors
+    ///
+    /// * If the content extends outside the memory allocated to the guest. In Wasmer, it returns `MemoryAccessViolation` error message;
+    /// * If `register_id` is pointing to unused register returns `InvalidRegisterId` error message.
+    ///
+    /// # Undefined Behavior
+    ///
+    /// If the content of register extends outside the preallocated memory on the host side, or the pointer points to a
+    /// wrong location this function will overwrite memory that it is not supposed to overwrite causing an undefined behavior.
+    pub fn read_register(&mut self, register_id: u64, ptr: u64) -> Result<()> {
+        let Self { registers, memory, .. } = self;
+        let register = registers.get(&register_id).ok_or(HostError::InvalidRegisterId)?;
+        Self::memory_set(*memory, ptr, register)
+    }
+
+    /// Returns the size of the blob stored in the given register.
+    /// * If register is used, then returns the size, which can potentially be zero;
+    /// * If register is not used, returns `u64::MAX`
+    ///
+    /// # Arguments
+    ///
+    /// * `register_id` -- a register id from where to read the data;
+    pub fn register_len(&mut self, register_id: u64) -> Result<u64> {
+        Ok(self.registers.get(&register_id).map(|r| r.len() as _).unwrap_or(std::u64::MAX))
+    }
+
+    /// Copies `data` into register. If register is unused will initialize it. If register has
+    /// larger capacity than needed for `data` will not re-allocate it. The register will lose
+    /// the pre-existing data if any.
+    ///
+    /// # Arguments
+    ///
+    /// * `register_id` -- a register into which to write the data;
+    /// * `data` -- data to be copied into register.
+    pub fn write_register(&mut self, register_id: u64, data: &[u8]) -> Result<()> {
+        let Self { registers, config, .. } = self;
+        Self::internal_write_register(registers, config, register_id, data)
+    }
+
+    fn internal_write_register(
```

Comment:
This function should probably charge gas relative to `data.len()` later.

Reply:
Will add proper gas charging for host functions later.